ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,372 @@
+    
+getDb_amk :: ( LMDB.Mode mode)
+  =>Tables.ApplyMapKind Tables.NamesMK k v
+  -> LMDB.Transaction mode (Tables.ApplyMapKind Tables.LMDBMK k v)
+getDb_amk (Tables.ApplyNamesMK name) = Tables.ApplyLMDBMK name <$>
+  LMDB.getDatabase (Just name)
+  
+initLMDBTable_amk :: (Serialise k, Serialise v)
+  => Tables.ApplyMapKind Tables.LMDBMK k v
+  -> Tables.ApplyMapKind Tables.ValuesMK k v
+  -> LMDB.Transaction ReadWrite (Tables.ApplyMapKind Tables.EmptyMK k v)
+initLMDBTable_amk (Tables.ApplyLMDBMK tbl_name db) (Tables.ApplyValuesMK vals) =
+  initLMDBTable tbl_name db vals $> Tables.ApplyEmptyMK 
+
+readLMDBTable_amk :: (Ord k, Serialise k, Serialise v)
+  => Tables.ApplyMapKind Tables.LMDBMK k v
+  -> Tables.ApplyMapKind Tables.KeysMK k v
+  -> LMDB.Transaction ReadWrite (Tables.ApplyMapKind Tables.ValuesMK k v)
+readLMDBTable_amk (Tables.ApplyLMDBMK _ db) (Tables.ApplyKeysMK keys) =
+  Tables.ApplyValuesMK <$> readLMDBTable db keys
+
+writeLMDBTable_amk :: (Serialise k, Serialise v)
+  => Tables.ApplyMapKind Tables.LMDBMK k v
+  -> Tables.ApplyMapKind Tables.DiffMK k v
+  -> LMDB.Transaction ReadWrite (Tables.ApplyMapKind Tables.EmptyMK k v)
+writeLMDBTable_amk (Tables.ApplyLMDBMK _tbl_name db) (Tables.ApplyDiffMK diff) =
+  writeLMDBTable db diff $> Tables.ApplyEmptyMK 
+
+readDbSettings :: LMDB.DBRef LMDB.ReadWrite DbState -> LMDB.Transaction mode DbState
+readDbSettings dbSettings = liftIO (LMDB.readDBRef dbSettings) >>= \case
+  Just x -> pure x
+  Nothing -> Exn.throw $ DbErrNoSettings
+
+withDbSettingsRW :: LMDB.DBRef LMDB.ReadWrite DbState -> (DbState -> LMDB.Transaction LMDB.ReadWrite (a, DbState)) -> LMDB.Transaction LMDB.ReadWrite a
+withDbSettingsRW dbSettings f  =
+  readDbSettings dbSettings >>= f >>= \(r, new_s) ->
+    liftIO $ LMDB.writeDBRef dbSettings (Just new_s) $> r
+
+readDbSettingsMaybeNull :: LMDB.DBRef LMDB.ReadWrite DbState -> LMDB.Transaction mode (Maybe DbState)
+readDbSettingsMaybeNull dbSettings = liftIO (LMDB.readDBRef dbSettings)
+
+withDbSettingsRWMaybeNull :: LMDB.DBRef LMDB.ReadWrite DbState -> (Maybe DbState -> LMDB.Transaction LMDB.ReadWrite (a, DbState)) -> LMDB.Transaction LMDB.ReadWrite a
+withDbSettingsRWMaybeNull dbSettings f  =
+  readDbSettingsMaybeNull dbSettings >>= f >>= \(r, new_s) ->
+    liftIO $ LMDB.writeDBRef dbSettings (Just new_s) $> r
+
+data GuardDbDir  = GDDMustExist | GDDMustNotExist
+
+guardDbDir :: IOLike m => GuardDbDir -> FS.SomeHasFS m -> FS.FsPath -> m FilePath
+guardDbDir gdd (FS.SomeHasFS fs) path = do
+  FS.doesFileExist fs path >>= \b -> when b $ Exn.throw $ DbErrStr $ "guardDbDir:must be a directory:" <> show path
+  FS.doesDirectoryExist fs path >>= \case
+    True | GDDMustNotExist <- gdd ->Exn.throw $ DbErrStr $ "guardDbDir: Must not exist: " <> show path
+    _ -> pure ()
+  FS.createDirectoryIfMissing fs True path
+  FS.unsafeToFilePath fs path
+
+defaultLMDBLimits :: LMDB.Limits
+defaultLMDBLimits = LMDB.defaultLimits
+  { LMDB.mapSize = 6_000_000_000
```

Comment:
We'll need to figure out how to maintain this `mapSize` 😬 Maybe the user should set it in a config file?

Reply:
Yes, the whole limits needs to be configurable. I had thought this could perhaps come in from a config file through the BackendSlector

Reply:
(I'm leaving this Conversation open as a TODO.)